pkg/service/providercacher/cachingqueuepoller.go

```diff
@@ -0,0 +1,191 @@
+}
+
+// Option configures the CachingQueuePoller
+type Option func(*CachingQueuePoller)
+
+// WithPollInterval sets the polling interval for the queue
+func WithPollInterval(interval time.Duration) Option {
+	return func(p *CachingQueuePoller) {
+		p.interval = interval
+	}
+}
+
+// WithJobBatchSize sets the maximum number of jobs to process in a batch
+func WithJobBatchSize(size int) Option {
+	return func(p *CachingQueuePoller) {
+		p.jobBatchSize = size
+	}
+}
+
+// NewCachingQueuePoller creates a new CachingQueuePoller instance.
+func NewCachingQueuePoller(queue CachingQueue, cacher ProviderCacher, opts ...Option) (*CachingQueuePoller, error) {
+	poller := &CachingQueuePoller{
+		queue:        queue,
+		cacher:       cacher,
+		interval:     defaultPollInterval,
+		jobBatchSize: defaultJobBatchSize,
+		stopped:      make(chan struct{}),
+	}
+
+	// Apply options
+	for _, opt := range opts {
+		opt(poller)
+	}
+
+	if poller.interval <= 0 {
+		return nil, fmt.Errorf("poll interval %v must be positive", poller.interval)
+	}
+
+	if poller.jobBatchSize > maxJobBatchSize {
+		return nil, fmt.Errorf("job batch size %d exceeds maximum allowed %d", poller.jobBatchSize, maxJobBatchSize)
+	}
+
+	return poller, nil
+}
+
+// Start begins polling the queue for caching jobs.
+func (p *CachingQueuePoller) Start() {
+	p.startOnce.Do(func() {
+		// Create root context
+		p.ctx, p.cancel = context.WithCancel(context.Background())
+		log.Info("Starting caching queue poller")
+
+		go func() {
+			timer := time.NewTimer(p.interval)
+
+			for {
+				select {
+				case <-timer.C:
+					p.processJobs(p.ctx)
+					timer.Reset(p.interval)
+
+				case <-p.ctx.Done():
+					log.Info("Stopping polling loop")
+					close(p.stopped)
+					return
+				}
+			}
+		}()
+	})
+}
+
+// Stop stops the polling loop and waits for it to finish.
+func (p *CachingQueuePoller) Stop() {
+	p.stopOnce.Do(func() {
+		// Cancel the root context, which will cancel all child contexts
+		if p.cancel != nil {
+			p.cancel()
+		}
+
+		// Wait for the polling loop to finish
+		<-p.stopped
+	})
+}
+
+// processJobs reads and processes all available jobs from the queue in batches
+func (p *CachingQueuePoller) processJobs(ctx context.Context) {
+	var wg sync.WaitGroup
+	maxGoroutines := make(chan struct{}, maxParallelBatches) // Semaphore to limit concurrent batches
+
+	// Keep processing batches until the queue is empty
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case maxGoroutines <- struct{}{}: // Acquire semaphore
+		}
+
+		// Read a batch of jobs
+		jobs, err := p.queue.ReadJobs(ctx, p.jobBatchSize)
+		if err != nil {
+			log.Errorf("Error reading jobs from queue: %v", err)
+			break
+		}
+
+		if len(jobs) == 0 {
+			// No more jobs in the queue
+			break
+		}
+
+		// Process batch in a new goroutine
+		wg.Add(1)
+		go func(batch []ProviderCachingJob) {
+			defer func() {
+				// Release semaphore when done
+				<-maxGoroutines
+				wg.Done()
+			}()
+
+			p.processBatch(ctx, batch)
```

Comment:
I think this will be fine, but you're still restricting the maximum concurrent jobs by waiting for a whole batch to finish. Personally I'd have something reading from the queue and putting individual jobs on a channel, and many channel consumers so you maintain the total number of concurrent tasks.

Right now if you have a slow job in every batch then you'll end up with just `maxParallelBatches` running jobs at times, whereas you really want to maintain `maxParallelBatches * jobBatchSize` at _all_ times.